### PR TITLE
Migrate to Quicknode for ipfs upload

### PR DESCRIPTION
**Pull request by AhsanRns**
## Description

Closes: #XXXX

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: easel/lib/easel_provider.dart [hunk 1]
```diff
@@ -814,12 +813,27 @@ class EaselProvider extends ChangeNotifier {
       uploadThumbnailResponse = uploadResponse.getOrElse(() => StorageResponseModel.initial());
     }
 
-    final response = await repository.uploadFile(
-      file: _file!,
-      onUploadProgressCallback: (value) {
-        _uploadProgressController.sink.add(value);
-      },
-    );
+    final whereToUpload = QuickNode.listOfQuickNodeAllowedExtension().contains(fileExtension.toLowerCase());
```
Comment (kjawadDeveloper2) [inline easel/lib/easel_provider.dart:816]: The name doesn't justifies what it is doing. Maybe you should use `shouldUploadToQuickNode`

File: easel/lib/easel_provider.dart [hunk 2]
```diff
@@ -814,12 +813,27 @@ class EaselProvider extends ChangeNotifier {
...
+    if (!whereToUpload) {
+      response = await repository.uploadFile(
+        file: _file!,
+        onUploadProgressCallback: (value) {
+          _uploadProgressController.sink.add(value);
+        },
+      );
+    } else {
+      response = await repository.uploadFileUsingQuickNode(
+        uploadIPFSInput: UploadIPFSInput(
```
Comment (kjawadDeveloper2) [inline easel/lib/easel_provider.dart:829]: Does quick node doesn't provides any support for the upload progress? Kindly check line 824.

File: easel/lib/easel_provider.dart [hunk 3]
```diff
@@ -814,12 +813,27 @@ class EaselProvider extends ChangeNotifier {
...
+          fileName: fileName,
+          filePath: file?.path ?? '',
```
Comment (kjawadDeveloper2) [inline easel/lib/easel_provider.dart:831]: If the file doesn't have a path then it shouldn't be uploaded.

File: easel/lib/models/storage_response_model.dart [hunk 4]
```diff
@@ -21,6 +23,26 @@ class StorageResponseModel {
...
+            uploadIPFSOutput.info?.size ?? '0',
+          ),
+          pin: Pin(
+            cid: uploadIPFSOutput.pin?.cid,
+            created: uploadIPFSOutput.created,
+            size: int.parse(
+              uploadIPFSOutput.info?.size ?? '0',
+            ),
+            status: uploadIPFSOutput.status,
+          )),
```
Comment (kjawadDeveloper2) [inline easel/lib/models/storage_response_model.dart:43]:
```suggestion
          ),),
```

File: easel/lib/screens/choose_format_screen.dart [hunk 5]
```diff
@@ -36,6 +36,9 @@ class _ChooseFormatScreenState extends State<ChooseFormatScreen> {
       return;
     }
 
+
+
+
```
Comment (kjawadDeveloper2) [inline easel/lib/screens/choose_format_screen.dart:41]: You can revert these spaces.

File: easel/lib/services/datasources/remote_datasource.dart [hunk 6]
```diff
@@ -21,19 +22,33 @@ abstract class RemoteDataSource {
...
+    required this.httpClient,
+    required this.analyticsHelper,
+    required this.quickNode,
+  });
+
+  @override
+  Future<StorageResponseModel> uploadFileUsingQuickNode({required UploadIPFSInput uploadIPFSInput}) async {
+    final response = await quickNode.uploadNewObjectToIPFS(uploadIPFSInput);
+    return response;
+  }
```
Comment (kjawadDeveloper2) [inline easel/lib/services/datasources/remote_datasource.dart:51]:
```suggestion
  @override
  Future<StorageResponseModel> uploadFileUsingQuickNode({required UploadIPFSInput uploadIPFSInput}) => 
 quickNode.uploadNewObjectToIPFS(uploadIPFSInput);
```

File: easel/lib/services/third_party_services/quick_node.dart [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+abstract class QuickNode {
+  /// Upload a new object to IPFS and pins it for permanent storage on the network.
+  /// [UploadIPFSInput] as an input
+  /// [UploadIPFSOutput] as an output
+  Future<StorageResponseModel> uploadNewObjectToIPFS(UploadIPFSInput uploadIPFSInput);
+
+  /// these are the list of extension required
+  static List<String> listOfQuickNodeAllowedExtension() {
+    return ['pdf', 'mpeg', 'jpg', 'png', 'plain'];
+  }
```
Comment (kjawadDeveloper2) [inline easel/lib/services/third_party_services/quick_node.dart:18]:
```suggestion
  static List<String> listOfQuickNodeAllowedExtension()  => ['pdf', 'mpeg', 'jpg', 'png', 'plain'];
```

File: easel/lib/services/third_party_services/quick_node.dart [hunk 8]
```diff
@@ -0,0 +1,131 @@
+import 'dart:convert';
+import 'dart:io';
+import 'package:easel_flutter/env.dart';
+import 'package:easel_flutter/models/storage_response_model.dart';
+
+///* using this because dio is already init with other  base url
+import 'package:http/http.dart' as http;
```
Comment (kjawadDeveloper2) [inline easel/lib/services/third_party_services/quick_node.dart:7]: As i shared in our chat you can re-use dio by assign the url only when the backend call is needed. Having dio and http in the code base not very pleasing.

File: easel/lib/services/third_party_services/quick_node.dart [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+    final fileStream = http.ByteStream(file.openRead());
+    final fileLength = await file.length();
+    final multipartFile = http.MultipartFile('Body', fileStream, fileLength, filename: file.path.split('/').last);
+
+    request.files.add(multipartFile);
+
+    final response = await http.Response.fromStream(await request.send());
+
+    final a = UploadIPFSOutput.fromJson(json.decode(response.body) as Map<String, dynamic>);
```
Comment (kjawadDeveloper2) [inline easel/lib/services/third_party_services/quick_node.dart:59]:
```suggestion
    final uploadIPFSOutput = UploadIPFSOutput.fromJson(json.decode(response.body) as Map<String, dynamic>);
```

File: easel/lib/services/third_party_services/quick_node.dart [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+  UploadIPFSInput({
+    required this.fileName,
+    required this.filePath,
+    required this.contentType,
+  });
+
+  Map<String, dynamic> toJson() {
+    return {};
+  }
```
Comment (kjawadDeveloper2) [inline easel/lib/services/third_party_services/quick_node.dart:78]: Is this method needed?

File: easel/lib/services/third_party_services/quick_node.dart [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        cid: json['cid'] as String?,
+        name: json['name'] as String?,
+      );
+}
+
+class Info {
+  final String? size;
+
+  Info({
+    this.size,
```
Comment (kjawadDeveloper2) [inline easel/lib/services/third_party_services/quick_node.dart:127]:
```suggestion
    this.size
```

File: easel/lib/services/third_party_services/quick_node.dart [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+class Pin {
+  String? cid;
+  String? name;
+
+  Pin({
+    this.cid,
+    this.name,
+  });
```
Comment (kjawadDeveloper2) [inline easel/lib/services/third_party_services/quick_node.dart:115]:
```suggestion
  Pin({
    this.cid,
    this.name
  });
```

File: easel/lib/utils/dependency_injection/dependency_injection_container.dart [hunk 13]
```diff
@@ -62,15 +63,16 @@ void _registerExternalDependencies() {
...
 }
 
 void _registerLocalDataSources() {
   sl.registerLazySingleton<LocalDataSource>(() => LocalDataSourceImpl(sharedPreferences: sl(), database: sl(), cacheManager: sl()));
 }
 
 void _registerProviders() {
-  sl.registerLazySingleton<EaselProvider>(() => EaselProvider(videoPlayerHelper: sl(), audioPlayerHelperForFile: sl(), fileUtilsHelper: sl(), repository: sl(), audioPlayerHelperForUrl: sl(), mediaInfo: sl()));
+  sl.registerLazySingleton<EaselProvider>(
+      () => EaselProvider(videoPlayerHelper: sl(), audioPlayerHelperForFile: sl(), fileUtilsHelper: sl(), repository: sl(), audioPlayerHelperForUrl: sl(), mediaInfo: sl()));
```
Comment (kjawadDeveloper2) [inline easel/lib/utils/dependency_injection/dependency_injection_container.dart:75]:
```suggestion
      () => EaselProvider(videoPlayerHelper: sl(), audioPlayerHelperForFile: sl(), fileUtilsHelper: sl(), repository: sl(), audioPlayerHelperForUrl: sl(), mediaInfo: sl(),));
```

**Review by kjawadDeveloper2 — COMMENTED** [review]
Have you checked after uploading the image that the wallet is showing the image/pdf/etc.

File: easel/lib/easel_provider.dart (hunk 1 above)
Comment (AhsanRns) [reply easel/lib/easel_provider.dart:816]: done

File: easel/lib/screens/choose_format_screen.dart (hunk 5 above)
Comment (AhsanRns) [reply easel/lib/screens/choose_format_screen.dart:41]: done

File: easel/lib/utils/dependency_injection/dependency_injection_container.dart (hunk 13 above)
Comment (AhsanRns) [reply easel/lib/utils/dependency_injection/dependency_injection_container.dart:75]: done

File: easel/lib/services/third_party_services/quick_node.dart (hunk 12 above)
Comment (AhsanRns) [reply easel/lib/services/third_party_services/quick_node.dart:115]: done

File: easel/lib/services/third_party_services/quick_node.dart (hunk 11 above)
Comment (AhsanRns) [reply easel/lib/services/third_party_services/quick_node.dart:127]: done

File: easel/lib/services/third_party_services/quick_node.dart (hunk 10 above)
Comment (AhsanRns) [reply easel/lib/services/third_party_services/quick_node.dart:78]: removed

File: easel/lib/services/third_party_services/quick_node.dart (hunk 9 above)
Comment (AhsanRns) [reply easel/lib/services/third_party_services/quick_node.dart:59]: done

File: easel/lib/models/storage_response_model.dart (hunk 4 above)
Comment (AhsanRns) [reply easel/lib/models/storage_response_model.dart:43]: done

File: easel/lib/easel_provider.dart (hunk 3 above)
Comment (AhsanRns) [reply easel/lib/easel_provider.dart:831]: done

File: easel/lib/easel_provider.dart (hunk 1 above)
Comment (AhsanRns) [reply easel/lib/easel_provider.dart:816]: done

File: easel/lib/easel_provider.dart (hunk 2 above)
Comment (AhsanRns) [reply easel/lib/easel_provider.dart:829]: yes it is supported working on it

**Review by kjawadDeveloper2 — APPROVED** [review]

**Pull request merged**